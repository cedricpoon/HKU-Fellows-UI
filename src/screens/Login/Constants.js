export const ALERT_DURATION = 3000;
