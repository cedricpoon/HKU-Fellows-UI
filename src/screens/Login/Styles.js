import { StyleSheet } from 'react-native';

import { logo } from 'hkufui/theme/palette';
import { fontSize } from 'hkufui/theme/grid'

export default StyleSheet.create({
  fullPageContainer: {
    flex: 1,
    justifyContent: 'center'
  },
  inputForm: {
    paddingHorizontal: 30,
    paddingBottom: 60,
    zIndex: 2
  },
  submit: {
    marginTop: 30,
  },
  declaration: {
    fontSize: fontSize.miniSize,
    fontStyle: 'italic'
  },
  placeholder: {
    color: logo.black
  },
  noBorderItem: {
    borderBottomWidth: 0,
    marginTop: 15
  },
  backdrop: {
    position: 'absolute',
    zIndex: 1
  }
});
