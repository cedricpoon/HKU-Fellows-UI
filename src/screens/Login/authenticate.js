import { LOGIN_ERROR, LOGGED_IN, LOGGING_IN } from 'hkufui/src/constants/actionTypes';
import { link, login } from 'hkufui/config/webapi';
import { localize } from 'hkufui/locale';

const locale = localize({ language: 'en', country: 'hk' });

export function onLogin({ username, password }, alert) {
  return async (dispatch) => {
    dispatch({ type: LOGGING_IN });
    try {
      let response = await fetch(link(login.password), {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ username, password }),
      });
      let res = await response.json();
      if (res.status === 200) {
        // successful return
        dispatch({
          type: LOGGED_IN,
          payload: {
            userId: username.toLowerCase(),
            ...res.payload
          }
        });
      } else {
        // failure
        let error = `${locale['login.errorMsgPrefix']} ${res.status}`;
        switch(res.status) {
          case 401:
            error = locale['login.failureMsg'];
            break;
          case 421:
            error = locale['login.laterMsg'];
            break;
        }
        alert(error);
        dispatch({ type: LOGIN_ERROR });
      }
    } catch (error) {
      alert(locale['fetch.noNetwork']);
      dispatch({ type: LOGIN_ERROR });
    }
  };
}

const handleExpandPosts = (state = {}, action = {}) => {
  switch (action.type) {
    case LOGGING_IN:
      return {}; // empty object as logging in
    case LOGIN_ERROR:
      return null; // clear credential
    case LOGGED_IN:
      return {
        ...state,
        ...action.payload
      }
    default:
      return state;
  }
};

export default handleExpandPosts;
