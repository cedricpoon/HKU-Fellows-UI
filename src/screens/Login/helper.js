export async function retrieveToken() {
  // Get the current token if it exists
  try {
    const firebase = require('react-native-firebase');

    const fcmToken = await firebase.messaging().getToken();
    return fcmToken
  } catch (e) {
    return null;
  }
}
