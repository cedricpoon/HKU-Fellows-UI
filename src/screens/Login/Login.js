import React, { Component } from 'react';
import { Container, Toast } from 'native-base';
import { withNavigation } from 'react-navigation';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { LoginForm } from 'hkufui/components';
import { localize } from 'hkufui/locale';
import { onLogin } from './authenticate';
import { ALERT_DURATION } from './Constants';

const locale = localize({ language: 'en', country: 'hk' });

const alert = (message) => {
  Toast.show({
    text: message,
    buttonText: locale['toast.dismiss'],
    type: 'danger',
    duration: ALERT_DURATION
  });
}

export class Login extends Component {
  constructor(props) {
    super(props);
    this._redirect = this._redirect.bind(this);
  }

  _redirect() {
    const { credential, navigation } = this.props;

    const loggedIn = credential && Object.keys(credential).length !== 0;
    // logged in page redirection
    if (loggedIn) {
      navigation.navigate('Context');
    }
  }

  componentDidMount() {
    this._redirect();
  }

  componentDidUpdate() {
    this._redirect();
  }

  render() {
    const { onLogin, credential } = this.props;

    const loggingIn = credential && Object.keys(credential).length === 0;

    return (
      <Container>
        <LoginForm
          alert={alert}
          onLogin={onLogin}
          loggingIn={loggingIn}
        />
      </Container>
    );
  }
}

Login.propTypes = {
  onLogin: PropTypes.func.isRequired,
  credential: PropTypes.object
}

const mapStateToProps = state => ({
  credential: state.credential
});

const mapDispatchToProps = dispatch => ({
  onLogin: ({ username, password }) => {dispatch(
    onLogin(
      { username, password },
      alert
    )
  )}
})

export default withNavigation(connect(
  mapStateToProps,
  mapDispatchToProps
)(Login));
