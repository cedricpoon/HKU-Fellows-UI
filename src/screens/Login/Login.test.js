import React from 'react';
import { shallow } from 'enzyme';
import { Login } from './Login';

describe('Testing Login Screen', () => {
  const requiredProps = {
    onLogin: () => {},
    onClearCredential: () => {}
  }

  it('renders as expected', () => {
    const wrapper = shallow(
      <Login {...requiredProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('has credential', () => {
    const wrapper = shallow(
      <Login {...requiredProps} credential={{}} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
