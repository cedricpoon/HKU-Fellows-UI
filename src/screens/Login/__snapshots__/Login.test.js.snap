// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login Screen has credential 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    credential={Object {}}
    onLogin={[Function]}
    onLogout={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <LoginForm
        alert={[Function]}
        loggingIn={true}
        onLogin={[Function]}
      />,
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "alert": [Function],
        "loggingIn": true,
        "onLogin": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <LoginForm
          alert={[Function]}
          loggingIn={true}
          onLogin={[Function]}
        />,
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "alert": [Function],
          "loggingIn": true,
          "onLogin": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing Login Screen renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    onLogin={[Function]}
    onLogout={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <LoginForm
        alert={[Function]}
        onLogin={[Function]}
      />,
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "alert": [Function],
        "loggingIn": undefined,
        "onLogin": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <LoginForm
          alert={[Function]}
          onLogin={[Function]}
        />,
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "alert": [Function],
          "loggingIn": undefined,
          "onLogin": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
